File: Network Delay Time - Leetcode 743/Network Delay Time - Leetcode 743.py
import heapq
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # Djikstra's Algorithm
        graph = defaultdict(list)
        for u, v, time in times:
            graph[u].append((v, time))
        
        min_times = {}
        min_heap = [(0, k)] # (distance from source to node, node)

        while min_heap:
            time_k_to_i, i = heapq.heappop(min_heap)
            if i in min_times:
                continue
            
            min_times[i] = time_k_to_i
            for nei, nei_time in graph[i]:
                if nei not in min_times:
                    heapq.heappush(min_heap, (time_k_to_i + nei_time, nei))
        
        if len(min_times) == n:
            return max(min_times.values())
        else:
            return -1
        # Time: O((V + E) log (V))
        # Space: O(V + E)

File: Network Delay Time - Leetcode 743/Network Delay Time - Leetcode 743.cpp
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using namespace std;

class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        unordered_map<int, vector<pair<int, int>>> graph;
        for (const auto& time : times) {
            int u = time[0], v = time[1], t = time[2];
            graph[u].emplace_back(v, t);
        }

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> minHeap;
        minHeap.emplace(0, k);
        unordered_map<int, int> minTimes;

        while (!minHeap.empty()) {
            auto [currentTime, node] = minHeap.top();
            minHeap.pop();

            if (minTimes.count(node)) continue;
            minTimes[node] = currentTime;

            for (const auto& [neighbor, time] : graph[node]) {
                if (!minTimes.count(neighbor)) {
                    minHeap.emplace(currentTime + time, neighbor);
                }
            }
        }

        if (minTimes.size() == n) {
            return max_element(minTimes.begin(), minTimes.end(),
                               [](const pair<int, int>& a, const pair<int, int>& b) {
                                   return a.second < b.second;
                               })->second;
        } else {
            return -1;
        }
    }
};
